### PR TITLE
Add Participation plugin

### PR DESCRIPTION
**Pull request by muXxer**

### CODE REVIEW

File: src/app/components/plugins/Participation.tsx [hunk 1]
```diff
@@ -0,0 +1,698 @@
...
+                        message: string;
+                    };
+                }>(
+                    url.origin,
+                    url.pathname,
+                    "get");
+
+                    return (response.data) ? response.data : response as IParticipationEventInfo;
+            } catch (err) {
+                console.log("Error fetch");
```
Comment (muXxer) [inline src/app/components/plugins/Participation.tsx:668]: should we leave the console logs here?

File: src/app/components/plugins/Participation.tsx [hunk 2]
```diff
@@ -0,0 +1,698 @@
...
+        }
+    }
+
+    /**
+     *  Submit event to be added.
+     */
+    private async submitEvent(): Promise<void> {
+        if (this.state.eventInfo) {
+            try {
+                const url = new URL(this.state.eventInfo);
```
Comment (muXxer) [inline src/app/components/plugins/Participation.tsx:524]: Should we maybe try to parse eventInfo as a valid JSON first, before doing requests?

File: src/models/plugins/participation/IParticipationEventQuestion.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import { IParticipationEventAnswer } from "./IParticipationEventAnswer";
+
+export interface IParticipationEventQuestion {
+    /**
+     * The text of the question
```
Comment (muXxer) [inline src/models/plugins/participation/IParticipationEventQuestion.ts:5]:
```suggestion
     * The text of the question.
```

File: src/models/plugins/participation/IParticipationEventQuestion.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+     */
+    text: string;
+
+    /**
+     * A possible answer for the question
```
Comment (muXxer) [inline src/models/plugins/participation/IParticipationEventQuestion.ts:10]:
```suggestion
     * The possible answers to the question.
```

File: src/models/plugins/participation/IParticipationEventPayload.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
+import { IParticipationEventQuestion } from "./IParticipationEventQuestion";
+
+export interface IParticipationEventPayload {
+    /**
+     * type of payload
```
Comment (muXxer) [inline src/models/plugins/participation/IParticipationEventPayload.ts:5]:
```suggestion
     * // The type of the event.
```

File: src/models/plugins/participation/IParticipationEventPayload.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+     */
+    type: number;
+
+    /**
+     *  Ballot questions
+     */
+    questions?: IParticipationEventQuestion[];
+
+    /**
+     * The description text of the staking event
```
Comment (muXxer) [inline src/models/plugins/participation/IParticipationEventPayload.ts:15]:
```suggestion
     * The description text of the participation event.
```

File: src/models/plugins/participation/IParticipationEventPayload.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+     */
+    denominator?: number;
+
+    /**
+     * The denominator used in the calculation of staking rewards.
+     */
+    requiredMinimumRewards?: number;
+
+    /**
+     * Additional information
```
Comment (muXxer) [inline src/models/plugins/participation/IParticipationEventPayload.ts:40]:
```suggestion
     * The additional description text about the participation event.
```

File: src/models/plugins/participation/IParticipationEventPayload.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+     */
+    numerator?: number;
+
+    /**
+     * The denominator used in the calculation of staking rewards.
+     */
+    denominator?: number;
+
+    /**
+     * The denominator used in the calculation of staking rewards.
```
Comment (muXxer) [inline src/models/plugins/participation/IParticipationEventPayload.ts:35]:
```suggestion
     * The minimum rewards required to be included in the staking results.
```

File: src/models/plugins/participation/IParticipationEventPayload.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+     */
+    symbol?: string;
+
+    /**
+     * The numerator used in the calculation of staking rewards.
+     */
+    numerator?: number;
+
+    /**
+     * The denominator used in the calculation of staking rewards.
```
Comment (muXxer) [inline src/models/plugins/participation/IParticipationEventPayload.ts:30]:
```suggestion
     * The denominator is used in combination with Numerator to calculate the rewards per milestone per staked tokens.
```

File: src/models/plugins/participation/IParticipationEventPayload.ts [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+     */
+    text?: string;
+
+    /**
+     * The symbol in UTF-8 format
+     */
+    symbol?: string;
+
+    /**
+     * The numerator used in the calculation of staking rewards.
```
Comment (muXxer) [inline src/models/plugins/participation/IParticipationEventPayload.ts:25]:
```suggestion
     * The numerator is used in combination with Denominator to calculate the rewards per milestone per staked tokens.
```

File: src/models/plugins/participation/IParticipationEventPayload.ts [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+     */
+    questions?: IParticipationEventQuestion[];
+
+    /**
+     * The description text of the staking event
+     */
+    text?: string;
+
+    /**
+     * The symbol in UTF-8 format
```
Comment (muXxer) [inline src/models/plugins/participation/IParticipationEventPayload.ts:20]:
```suggestion
     * The symbol of the rewarded token.
```

File: src/models/plugins/participation/IParticipationEventPayload.ts [hunk 12]
```diff
@@ -0,0 +1,43 @@
+import { IParticipationEventQuestion } from "./IParticipationEventQuestion";
+
+export interface IParticipationEventPayload {
+    /**
+     * type of payload
+     */
+    type: number;
+
+    /**
+     *  Ballot questions
```
Comment (muXxer) [inline src/models/plugins/participation/IParticipationEventPayload.ts:10]:
```suggestion
     *  The questions of the ballot and their possible answers.
```

File: src/models/plugins/participation/IParticipationEventInfo.ts [hunk 13]
```diff
@@ -0,0 +1,40 @@
+import { IParticipationEventPayload } from "./IParticipationEventPayload";
+import { IParticipationEventStatus } from "./IParticipationEventStatus";
+
+export interface IParticipationEventInfo {
+    /**
+     * The Event name
```
Comment (muXxer) [inline src/models/plugins/participation/IParticipationEventInfo.ts:6]:
```suggestion
     * The name of the event.
```

File: src/models/plugins/participation/IParticipationEventInfo.ts [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+     */
+    name: string;
+
+    /**
+     * The milestone where the event commences.
```
Comment (muXxer) [inline src/models/plugins/participation/IParticipationEventInfo.ts:11]:
```suggestion
     * The milestone index the commencing period starts.
```

File: src/models/plugins/participation/IParticipationEventInfo.ts [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+     */
+    milestoneIndexCommence: string;
+
+    /**
+     * The milestone where the event enters holding phase.
```
Comment (muXxer) [inline src/models/plugins/participation/IParticipationEventInfo.ts:16]:
```suggestion
     * The milestone index the holding period starts.
```

File: src/models/plugins/participation/IParticipationEventInfo.ts [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+     */
+    milestoneIndexStart: string;
+
+    /**
+     * The milestone where the event ends.
```
Comment (muXxer) [inline src/models/plugins/participation/IParticipationEventInfo.ts:21]:
```suggestion
     * The milestone index the event ends.
```

File: src/models/plugins/participation/IParticipationEventInfo.ts [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+     */
+    milestoneIndexEnd: string;
+
+    /**
+     * The event info payload.
```
Comment (muXxer) [inline src/models/plugins/participation/IParticipationEventInfo.ts:26]:
```suggestion
     * The payload of the event (ballot/staking).
```

File: src/models/plugins/participation/IParticipationEventInfo.ts [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+     */
+    payload: IParticipationEventPayload;
+
+    /**
+     * Additional information.
```
Comment (muXxer) [inline src/models/plugins/participation/IParticipationEventInfo.ts:31]:
```suggestion
     * The additional description text about the event.
```

File: src/models/plugins/participation/IParticipationEventInfo.ts [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+
+    /**
+     * The milestone where the event enters holding phase.
+     */
+    milestoneIndexStart: string;
+
+    /**
+     * The milestone where the event ends.
+     */
+    milestoneIndexEnd: string;
```
Comment (muXxer) [inline src/models/plugins/participation/IParticipationEventInfo.ts:23]: `milestoneIndexEnd` is a number!

File: src/models/plugins/participation/IParticipationEventInfo.ts [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+
+    /**
+     * The milestone where the event commences.
+     */
+    milestoneIndexCommence: string;
+
+    /**
+     * The milestone where the event enters holding phase.
+     */
+    milestoneIndexStart: string;
```
Comment (muXxer) [inline src/models/plugins/participation/IParticipationEventInfo.ts:18]: `milestoneIndexStart` is a number!

File: src/models/plugins/participation/IParticipationEventInfo.ts [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+export interface IParticipationEventInfo {
+    /**
+     * The Event name
+     */
+    name: string;
+
+    /**
+     * The milestone where the event commences.
+     */
+    milestoneIndexCommence: string;
```
Comment (muXxer) [inline src/models/plugins/participation/IParticipationEventInfo.ts:13]: `milestoneIndexCommence` is a number!

File: src/models/plugins/participation/IParticipationEventStatus.ts [hunk 22]
```diff
@@ -0,0 +1,27 @@
+export interface IParticipationEventStatus {
+    /**
+     * Status of the event
+     */
+    status: string;
+
+    /**
+     * The milestone that confirmed it.
+     */
+    milestoneIndex: string;
```
[the remark this answers is not in the document]
Comment (brancoder) [reply src/models/plugins/participation/IParticipationEventStatus.ts]: Changed type to number

File: src/models/plugins/participation/IParticipationEventStaking.ts [hunk 23]
```diff
@@ -0,0 +1,16 @@
+export interface IParticipationEventStaking {
```
[the remark this answers is not in the document]
Comment (brancoder) [reply src/models/plugins/participation/IParticipationEventStaking.ts]: Deleted unused file.

File: src/app/components/plugins/Participation.tsx [hunk 24]
```diff
@@ -0,0 +1,698 @@
...
+                                        </div>
+                                    )}
+                                </div>
+                            )}
+                            {this.state.dialogType === "add" && (
+                                <React.Fragment>
+                                    <p>Please enter the details of the event to {this.state.dialogType}.</p>
+
+                                    <div className="dialog--label">
+                                        Json Config or URL
```
[the remark this answers is not in the document]
Comment (brancoder) [reply src/app/components/plugins/Participation.tsx]: Changed Config to Configuration.

File: src/app/components/plugins/Participation.tsx [hunk 25]
```diff
@@ -0,0 +1,698 @@
...
+                                            className="textarea--stretch
+                                            dialog--value-textarea
+                                            dialog--value-textarea__json"
+                                            placeholder='e.g. { "name":"Example event title"... } or http://example.com'
+                                            value={this.state.eventInfo}
+                                            disabled={this.state.dialogBusy}
+                                            onChange={e => this.setState({ eventInfo: e.target.value })}
+                                        />
+                                        {this.state.eventInfo && !this.validateJsonOrUrl(this.state.eventInfo) && (
+                                        <span className="validation--error">Not a valid JSON or Url</span>
```
[the remark this answers is not in the document]
Comment (brancoder) [reply src/app/components/plugins/Participation.tsx]: Changed to URL

File: src/app/components/plugins/Participation.tsx (hunk 1 above)
Comment (brancoder) [reply src/app/components/plugins/Participation.tsx]: Removed console log.

File: src/app/components/plugins/Participation.tsx (hunk 2 above)
Comment (brancoder) [reply src/app/components/plugins/Participation.tsx]: Changed to parse eventInfo as valid JSON first.

File: src/models/plugins/participation/IParticipationEventQuestion.ts [hunk 26]
```diff
@@ -0,0 +1,18 @@
...
+     */
+    answers: IParticipationEventAnswer[];
+
+    /**
+     * Additional information about the question
+     */
+    additionalInfo: number;
```
[the remark this answers is not in the document]
Comment (brancoder) [reply src/models/plugins/participation/IParticipationEventQuestion.ts]: Changed type to string.

File: src/models/plugins/participation/IParticipationEventInfo.ts (hunk 19 above)
Comment (brancoder) [reply src/models/plugins/participation/IParticipationEventInfo.ts]: Changed type to number.

File: src/models/plugins/participation/IParticipationEventInfo.ts (hunk 20 above)
Comment (brancoder) [reply src/models/plugins/participation/IParticipationEventInfo.ts]: Changed type to number.

File: src/models/plugins/participation/IParticipationEventInfo.ts (hunk 21 above)
Comment (brancoder) [reply src/models/plugins/participation/IParticipationEventInfo.ts]: Changed type to number.

File: src/app/routes/Settings.tsx [hunk 27]
```diff
@@ -62,6 +63,10 @@ class Settings extends AsyncComponent<unknown, SettingsState> {
             if (pluginDetails) {
```
[the remark this answers is not in the document]
Comment (brancoder) [reply src/app/routes/Settings.tsx]: Renamed to pluginDetailsSpammer and pluginDetailsParticipation.

**Review by brancoder — DISMISSED** [review]

**Review by brancoder — APPROVED** [review]

**Pull request merged**